generators/tracegen/src/main/scala/Tile.scala

```diff
@@ -29,6 +34,190 @@ class TraceGenTile(val id: Int, val params: TraceGenParams, q: Parameters)
   override lazy val module = new TraceGenTileModuleImp(this)
 }
 
+class BoomLSUShim(implicit p: Parameters) extends BoomModule()(p)
+  with MemoryOpConstants {
+  val io = IO(new Bundle {
+    val lsu = Flipped(new LSUCoreIO)
+    val tracegen = Flipped(new HellaCacheIO)
+  })
+
+  io.lsu.tsc_reg := 0.U(1.W)
+
+  val rob_sz = numRobEntries
+  val rob = Reg(Vec(rob_sz, new HellaCacheReq))
+  val rob_respd = RegInit(VecInit((~(0.U(rob_sz.W))).asBools))
+  val rob_uop = Reg(Vec(rob_sz, new MicroOp))
+  val rob_bsy  = RegInit(VecInit(0.U(rob_sz.W).asBools))
+  val rob_head = RegInit(0.U(log2Up(rob_sz).W))
+  val rob_tail = RegInit(0.U(log2Up(rob_sz).W))
+  val rob_wait_till_empty = RegInit(false.B)
+  val ready_for_amo = rob_tail === rob_head && io.lsu.fencei_rdy
+  when (ready_for_amo) {
+    rob_wait_till_empty := false.B
+  }
+
+  def WrapInc(idx: UInt, max: Int): UInt = {
+    Mux(idx === (max-1).U, 0.U, idx + 1.U)
+  }
+
+
+  io.tracegen.req.ready := (!rob_bsy(rob_tail) &&
+    !rob_wait_till_empty &&
+    (ready_for_amo || !(isAMO(io.tracegen.req.bits.cmd) || io.tracegen.req.bits.cmd === M_XLR || io.tracegen.req.bits.cmd === M_XSC)) &&
+    (WrapInc(rob_tail, rob_sz) =/= rob_head) &&
+    !(io.lsu.ldq_full(0) && isRead(io.tracegen.req.bits.cmd)) &&
+    !(io.lsu.stq_full(0) && isWrite(io.tracegen.req.bits.cmd))
+  )
+
+  val tracegen_uop = WireInit((0.U).asTypeOf(new MicroOp))
+  tracegen_uop.uses_ldq     := isRead(io.tracegen.req.bits.cmd) && !isWrite(io.tracegen.req.bits.cmd)
+  tracegen_uop.uses_stq     := isWrite(io.tracegen.req.bits.cmd)
+  tracegen_uop.rob_idx      := rob_tail
+  tracegen_uop.uopc         := io.tracegen.req.bits.tag
+  tracegen_uop.mem_size     := io.tracegen.req.bits.size
+  tracegen_uop.mem_cmd      := io.tracegen.req.bits.cmd
+  tracegen_uop.mem_signed   := io.tracegen.req.bits.signed
+  tracegen_uop.ldq_idx      := io.lsu.dis_ldq_idx(0)
+  tracegen_uop.stq_idx      := io.lsu.dis_stq_idx(0)
+  tracegen_uop.is_amo       := isAMO(io.tracegen.req.bits.cmd) || io.tracegen.req.bits.cmd === M_XSC
+  tracegen_uop.ctrl.is_load := isRead(io.tracegen.req.bits.cmd) && !isWrite(io.tracegen.req.bits.cmd)
+  tracegen_uop.ctrl.is_sta  := isWrite(io.tracegen.req.bits.cmd)
+  tracegen_uop.ctrl.is_std  := isWrite(io.tracegen.req.bits.cmd)
+
+  io.lsu.dis_uops(0).valid         := io.tracegen.req.fire()
+  io.lsu.dis_uops(0).bits          := tracegen_uop
+  
+  when (io.tracegen.req.fire()) {
+    rob_tail := WrapInc(rob_tail, rob_sz)
+    rob_bsy(rob_tail)   := true.B
```

Comment:
Aren't rob_head and rob_tail backwards here? Entries are placed into a circular buffer at the head and taken out at the tail.

Reply:
This follow's BOOM's ROB naming scheme.